File: frontend/src/types/events.ts
// 后端 emit 的 output 事件 payload
// workflow 的 output 事件复用同一结构，但 stream 额外允许协议帧 step-start/step-done
export interface OutputEventData {
  stream:
    | "stdout"
    | "stderr"
    | "llm"
    | "llm-thinking"
    | "llm-tool"
    | "logcat"
    | "logcat-replace"
    | "step-start"
    | "step-done"
    | "step-skip"
    | "progress";
  line: string;
  // workflow output 事件专有：该行归属的 step 索引（后端下发，见 api.executeWorkflow）。
  // 有它就按索引落桶，规避 Wails 事件乱序把 100% 折进下一个 step。action 事件无此字段。
  step?: string;
  // action output 事件专有：后端 emit 时的自增序号（见 api.execute）。Wails Event.Emit
  // 每次都各起一个 goroutine 投递，到达前端的顺序无保证；用 seq 在应用前重排还原真实产出顺序。
  seq?: number;
}

// stream="logcat" 的 line 是一段紧凑 JSON，对应后端 logcatPayload。
export interface LogcatEntry {
  date: string; // "08-08"；未解析行为 ""
  time: string; // "11:22:33.456"；未解析行为 ""
  pid: number;
  tid: number;
  level: string; // V/D/I/W/E/F
  tag: string;
  message: string;
}

// ——— 统一过滤规则（与 Go internal/adb/logcat/rule.go 逐字段对齐，json tag 即协议） ———
// 后端是唯一求值器：前端只编辑/下发规则，不再自带过滤逻辑（spec 2026-08-18）。

export type LogcatKey = "tag" | "message" | "pid" | "tid" | "any";
export type LogcatOp = "contains" | "exact" | "regex";

export interface LogcatToken {
  key: LogcatKey;
  op: LogcatOp;
  negated: boolean;
  value: string;
  // 前端 UI 专有：输入框未固化尾巴（草稿态）。下发后端时携带无害（后端不看此字段）。
  draft?: boolean;
}

export interface LogcatRule {
  tokens: LogcatToken[];
  minLevel: string; // V/D/I/W/E/F 阈值，下拉唯一入口；空 = V
  package: string; // 包名；空 = 全量进程（只读 chip，preset 带入 / × 移除）
}

export const EMPTY_LOGCAT_RULE: LogcatRule = { tokens: [], minLevel: "V", package: "" };

// stream="logcat-replace" 的 line 是一段 JSON（对应后端 replaceFrame）。
// 规则变更/清空后后端重筛 raw ring 整体重发：head=true 为首帧，收到即清空本地缓冲
// 整体替换，并读取 matched/total 与 tag 直方图；后续 chunk 只追加 entries。
export interface LogcatReplaceFrame {
  head: boolean;
  entries: LogcatEntry[];
  matched?: number;
  total?: number;
  tagHistogram?: Record<string, number>;
}

// LLM 会话终点读数（done 事件附带，LLM 动作专有；后端 api.llmReadout 构造）
export interface LlmReadoutData {
  durationMs?: number;
  costUsd?: number;
  inputTokens?: number;
  outputTokens?: number;
  // claude 会话 id（stream-json 的 session_id）：历史查看态顶部展示，可用于 --resume
  sessionId?: string;
}

// 后端 emit 的 done 事件 payload
export interface DoneEventData {
  exitCode: number;
  err: string;
  duration: string;
  readout?: LlmReadoutData;
  // action done 事件专有：延续 output 事件的 seq 序号空间（见 api.emitDoneSeq）。
  // done 事件同样各起独立 goroutine 投递，可能抢跑在还没到达的 output 行前面；
  // 用它接入 outputFold 的重排队列，确保退出码行落在所有 output 行之后。
  seq?: number;
}

// workflow 单个步骤的运行状态（前端根据协议帧维护）
export interface WorkflowStepState {
  index: number;
  status: "pending" | "running" | "done" | "error" | "skipped";
  exitCode?: number;
  lines: string[];
  // 上一行是否是 progress——决定下一条 progress 覆盖还是追加（同 action 的 lines 语义）
  lastWasProgress?: boolean;
}

File: internal/api/events.go
package api

import (
	"fmt"
	"strconv"
	"sync/atomic"
	"time"

	"github.com/wailsapp/wails/v3/pkg/application"

	"workflow-tool/internal/runner"
)

// 本文件是 action:<id>:* / workflow:<id>:* 事件契约的唯一权威定义：
// 事件名、payload 形状、到达顺序协议（seq 排序 / step 归桶）全部集中在这里，
// 改事件协议只需改本文件（+前端对应处理）。

// actionEvents 封装单个 action 的事件通道（action:<id>:output / action:<id>:done）。
//
// 顺序协议（前端契约）：Wails 的 Event.Emit 每次调用都各自起一个 goroutine 投递
// （application/events.go），事件到达前端的顺序无保证。action 是单一输出桶，
// 每条 output 自带递增 seq，done 事件带 seq+1 延续同一序号空间，前端按 seq 排序
// 还原真实产出顺序——done 若抢跑到还没到达的 output 前面，退出码行会被插错位置。
type actionEvents struct {
	app *application.App
	id  string
	seq atomic.Int64
}

func newActionEvents(app *application.App, id string) *actionEvents {
	return &actionEvents{app: app, id: id}
}

// nextSeq 取下一个事件序号。必须原子：emit 并非总是串行回调——logcat-stream 从
// flush ticker 协程、控制协程（规则拒绝告警）、pidRefresher（包未运行告警）三处
// 并发 emit（internal/adb/logcat/stream.go:90/189/408），后两处在 core.mu 之外。
// 前端 seqGate 依赖 seq 唯一有序来定 logcat-replace 帧与增量帧的先后。
func (e *actionEvents) nextSeq() int64 { return e.seq.Add(1) }

// EmitFunc 返回 runner 用的 emit：每行 output 自带递增 seq。
func (e *actionEvents) EmitFunc() runner.EmitFunc {
	return func(stream, line string) {
		e.app.Event.Emit(eventName(e.id, "output"), map[string]any{
			"stream": stream, "line": line, "seq": e.nextSeq(),
		})
	}
}

// Done 发送结束事件，seq 接续最后一条 output（seq+1），前端按序应用退出码行。
func (e *actionEvents) Done(exitCode int, errMsg string, d time.Duration, readout map[string]any) {
	e.emitDone(exitCode, errMsg, d, readout, e.seq.Load()+1)
}

// DoneUnordered 发送不参与排序的结束事件（payload 不带 seq，前端直接应用）。
// 用于 runner 启动前早退的场景（如工作目录校验失败），此时尚无 output 序号空间。
func (e *actionEvents) DoneUnordered(exitCode int, errMsg string, d time.Duration) {
	e.emitDone(exitCode, errMsg, d, nil, 0)
}

func (e *actionEvents) emitDone(exitCode int, errMsg string, d time.Duration, readout map[string]any, seq int64) {
	payload := map[string]any{
		"exitCode": exitCode,
		"err":      errMsg,
		"duration": d.String(),
	}
	if readout != nil {
		payload["readout"] = readout
	}
	if seq > 0 {
		payload["seq"] = seq
	}
	e.app.Event.Emit(eventName(e.id, "done"), payload)
}

// workflowEvents 封装单个 workflow 的事件通道（workflow:<id>:output / workflow:<id>:done）。
//
// 归桶协议（前端契约）：Emit 到达顺序无保证——push 末尾的 100% progress 会与
// step-done/下一个 step-start 抢跑，输了就被前端折进下一个 step 的输出里。
// 每条 output 事件带 step 归属（当前 step 索引），前端按索引落桶，与到达顺序无关。
//
// 并发安全：curStep 写发生在 executor goroutine 的 step 边界，读发生在 runner 的
// 输出 goroutine；runner.Run 返回前已 join 全部输出 goroutine，故两者有序不重叠。
type workflowEvents struct {
	app     *application.App
	id      string
	curStep string
}

func newWorkflowEvents(app *application.App, id string) *workflowEvents {
	return &workflowEvents{app: app, id: id}
}

// EmitFunc 返回 runner/executor 用的 emit：step-start/step-skip 更新当前 step 归属。
func (e *workflowEvents) EmitFunc() runner.EmitFunc {
	return func(stream, line string) {
		if stream == "step-start" || stream == "step-skip" {
			e.curStep = line
		}
		e.app.Event.Emit(workflowEventName(e.id, "output"), map[string]string{
			"stream": stream, "line": line, "step": e.curStep,
		})
	}
}

// Done 发送 workflow 结束事件（workflow 无 output 排序协议，done 直接应用）。
func (e *workflowEvents) Done(exitCode int, errMsg string, d time.Duration) {
	e.app.Event.Emit(workflowEventName(e.id, "done"), map[string]any{
		"exitCode": exitCode,
		"err":      errMsg,
		"duration": d.String(),
	})
}

func eventName(id, suffix string) string {
	return fmt.Sprintf("action:%s:%s", id, suffix)
}

func workflowEventName(id, suffix string) string {
	return fmt.Sprintf("workflow:%s:%s", id, suffix)
}

func errStr(err error) string {
	if err == nil {
		return ""
	}
	return err.Error()
}

// llmReadout 从 LLM Result.Outputs 挑出终点读数字段（数值化）与会话 id；无可读字段时返回 nil。
// cost/tokens 来自 stream-json 的 result 事件（runner.recordStructuredFields 写入 Outputs），
// duration 用后端精确计时；读数行随 done 事件下发（聊天页终点展示）。
// sessionId 单独也撑起 readout：失败/超时杀进程时 result 事件不会到达（无 cost/tokens），
// 但 system/init 的 session_id 已写入——此刻恰恰是用户最需要会话 id（查现场/--resume）的时候。
func llmReadout(outputs map[string]string, d time.Duration) map[string]any {
	r := map[string]any{"durationMs": d.Milliseconds()}
	// 会话 id（system/init 事件由 runner.recordStructuredFields 写入）：聊天页历史查看态顶部展示
	if sid := outputs["session_id"]; sid != "" {
		r["sessionId"] = sid
	}
	if v, err := strconv.ParseFloat(outputs["cost_usd"], 64); err == nil {
		r["costUsd"] = v
	}
	in, errIn := strconv.Atoi(outputs["input_tokens"])
	out, errOut := strconv.Atoi(outputs["output_tokens"])
	if errIn == nil && errOut == nil {
		r["inputTokens"] = in
		r["outputTokens"] = out
	}
	// 仅当既无成本读数也无会话 id 时才视为无可读字段（durationMs 单独不算，
	// done 事件 payload 本身已带 duration 字符串）。
	if _, ok := r["sessionId"]; !ok {
		if _, ok := r["costUsd"]; !ok {
			if _, ok := r["inputTokens"]; !ok {
				return nil
			}
		}
	}
	return r
}
